I ran a local code review and found 2 blocking issues.

1. **HIGH – `autorag_research/pipelines/generation/self_rag.py:191-192`**
   `_parse_reflection()` coerces JSON values with `bool(payload.get(...))`. If the model returns quoted booleans (for example `"should_retrieve": "false"`), Python treats those non-empty strings as truthy, so the pipeline incorrectly sets `action='RETRIEVE'` and `supported=True`.

   **Real repro:**
   ```bash
   uv run python - <<'PY'
   from autorag_research.pipelines.generation.self_rag import SelfRAGPipeline
   pipeline = SelfRAGPipeline.__new__(SelfRAGPipeline)
   print(pipeline._parse_reflection('{"should_retrieve": "false", "is_supported": "false", "critique": "Need revision"}'))
   PY
   ```
   Output:
   ```python
   {'action': 'RETRIEVE', 'supported': True, 'search_query': '', 'critique': 'Need revision'}
   ```
   That is the opposite of the intended control flow, and the current tests only cover real JSON booleans.

2. **HIGH – `tests/autorag_research/orm/repository/test_image_chunk_retrieved_result.py:73`**
   The new expected count was changed from `3` to `2`, but the seeded DB already contains `(query_id=4, pipeline_id=2)` in `postgresql/db/init/002-seed.sql:120-123`. After inserting the two test rows, `get_by_query_and_pipeline([2, 4], 2)` still returns 3 rows from a clean database, so this updated test now fails.

   **Real result:**
   ```bash
   uv run pytest -q tests/autorag_research/orm/repository/test_image_chunk_retrieved_result.py::test_bulk_insert
   ```
   Fails with:
   ```
   E       assert 3 == 2
   E        +  where 3 = len(...)
   ```
   `origin/dev` still expects `3` at this assertion, which matches the clean seeded state.

**Verification I ran**
- `make clean-docker && make docker-up && make docker-wait` ✅
- `uv run pytest -q tests/autorag_research/pipelines/generation/test_self_rag_pipeline.py` ✅ (`5 passed in 2.50s`)
- `uv run pytest -q tests/autorag_research/orm/service/test_generation_pipeline.py tests/autorag_research/orm/repository/test_image_chunk_retrieved_result.py` ❌ (`1 failed, 15 passed`)

<!-- dani:stage=review_round;job=2f82dc3db5004897ae6bf0cd98d02406;pr=331;round=1;issue=87 -->

--- a/autorag_research/pipelines/generation/self_rag.py
+++ b/autorag_research/pipelines/generation/self_rag.py
@@ -155,6 +155,20 @@ class SelfRAGPipeline(BaseGenerationPipeline):
             critique=critique or "Improve grounding and specificity.",
         )
 
+    def _coerce_reflection_flag(self, value: Any, default: bool = False) -> bool:
+        if isinstance(value, bool):
+            return value
+        if isinstance(value, str):
+            normalized = value.strip().lower()
+            if normalized in {"true", "yes", "1", "supported"}:
+                return True
+            if normalized in {"false", "no", "0", "unsupported", ""}:
+                return False
+            return default
+        if value is None:
+            return default
+        return bool(value)
+
     def _parse_key_value_reflection(self, response_text: str) -> dict[str, Any]:
         parsed: dict[str, Any] = {
             "action": "FINISH",
@@ -175,7 +189,7 @@ class SelfRAGPipeline(BaseGenerationPipeline):
                 if normalized_action in {"FINISH", "RETRIEVE", "REVISE"}:
                     parsed["action"] = normalized_action
             elif key == "SUPPORTED":
-                parsed["supported"] = value.lower() in {"yes", "true", "supported"}
+                parsed["supported"] = self._coerce_reflection_flag(value)
             elif key in {"SEARCH_QUERY", "FOLLOW_UP_QUERY"}:
                 parsed["search_query"] = value
             elif key == "CRITIQUE" and value:
@@ -188,8 +202,8 @@ class SelfRAGPipeline(BaseGenerationPipeline):
         if stripped.startswith("{"):
             try:
                 payload = json.loads(stripped)
-                should_retrieve = bool(payload.get("should_retrieve", False))
-                supported = bool(payload.get("is_supported", False))
+                should_retrieve = self._coerce_reflection_flag(payload.get("should_retrieve", False))
+                supported = self._coerce_reflection_flag(payload.get("is_supported", False))
                 if should_retrieve:
                     action = "RETRIEVE"
                 elif supported:

--- a/autorag_research/pipelines/generation/self_rag.py
+++ b/autorag_research/pipelines/generation/self_rag.py
@@ -169,6 +169,13 @@ class SelfRAGPipeline(BaseGenerationPipeline):
             return default
         return bool(value)
 
+    def _resolve_reflection_action(self, should_retrieve: bool, supported: bool) -> str:
+        if should_retrieve:
+            return "RETRIEVE"
+        if supported:
+            return "FINISH"
+        return "REVISE"
+
     def _parse_key_value_reflection(self, response_text: str) -> dict[str, Any]:
         parsed: dict[str, Any] = {
             "action": "FINISH",
@@ -176,6 +183,8 @@ class SelfRAGPipeline(BaseGenerationPipeline):
             "search_query": "",
             "critique": response_text.strip(),
         }
+        should_retrieve_flag: bool | None = None
+        supported_flag: bool | None = None
 
         for line in response_text.splitlines():
             if ":" not in line:
@@ -188,12 +197,21 @@ class SelfRAGPipeline(BaseGenerationPipeline):
                 normalized_action = value.upper()
                 if normalized_action in {"FINISH", "RETRIEVE", "REVISE"}:
                     parsed["action"] = normalized_action
-            elif key == "SUPPORTED":
-                parsed["supported"] = self._coerce_reflection_flag(value)
+            elif key == "SHOULD_RETRIEVE":
+                should_retrieve_flag = self._coerce_reflection_flag(value)
+            elif key in {"SUPPORTED", "IS_SUPPORTED"}:
+                supported_flag = self._coerce_reflection_flag(value)
+                parsed["supported"] = supported_flag
             elif key in {"SEARCH_QUERY", "FOLLOW_UP_QUERY"}:
                 parsed["search_query"] = value
             elif key == "CRITIQUE" and value:
                 parsed["critique"] = value
+
+        if should_retrieve_flag is not None or supported_flag is not None:
+            parsed["action"] = self._resolve_reflection_action(
+                should_retrieve=should_retrieve_flag or False,
+                supported=supported_flag or False,
+            )
 
         return parsed
 
@@ -204,14 +222,8 @@ class SelfRAGPipeline(BaseGenerationPipeline):
                 payload = json.loads(stripped)
                 should_retrieve = self._coerce_reflection_flag(payload.get("should_retrieve", False))
                 supported = self._coerce_reflection_flag(payload.get("is_supported", False))
-                if should_retrieve:
-                    action = "RETRIEVE"
-                elif supported:
-                    action = "FINISH"
-                else:
-                    action = "REVISE"
                 return {
-                    "action": action,
+                    "action": self._resolve_reflection_action(should_retrieve, supported),
                     "supported": supported,
                     "search_query": payload.get("follow_up_query", "") or payload.get("search_query", ""),
                     "critique": payload.get("critique", stripped),

--- a/tests/autorag_research/orm/repository/test_image_chunk_retrieved_result.py
+++ b/tests/autorag_research/orm/repository/test_image_chunk_retrieved_result.py
@@ -59,6 +59,7 @@ def test_bulk_insert(
     image_chunk_retrieved_result_repository: ImageChunkRetrievedResultRepository,
     db_session: Session,
 ):
+    inserted_keys = {(2, 2, 2), (4, 2, 1)}
     results_to_insert = [
         {"query_id": 4, "pipeline_id": 2, "image_chunk_id": 1, "rel_score": 0.75},
         {"query_id": 2, "pipeline_id": 2, "image_chunk_id": 2, "rel_score": 0.65},
@@ -70,9 +71,13 @@ def test_bulk_insert(
     assert inserted_count == 2
 
     results = image_chunk_retrieved_result_repository.get_by_query_and_pipeline([2, 4], 2)
-    assert len(results) == 2
+    assert len(results) == 3
+    assert {(result.query_id, result.pipeline_id, result.image_chunk_id) for result in results} == inserted_keys | {
+        (4, 2, 2)
+    }
 
     # Cleanup
-    image_chunk_retrieved_result_repository.delete_by_query_and_pipeline(4, 2)
-    image_chunk_retrieved_result_repository.delete_by_query_and_pipeline(2, 2)
+    for result in results:
+        if (result.query_id, result.pipeline_id, result.image_chunk_id) in inserted_keys:
+            db_session.delete(result)
     db_session.commit()

--- a/tests/autorag_research/orm/repository/test_image_chunk_retrieved_result.py
+++ b/tests/autorag_research/orm/repository/test_image_chunk_retrieved_result.py
@@ -2,7 +2,7 @@ import pytest
 from sqlalchemy.orm import Session
 
 from autorag_research.orm.repository.image_chunk_retrieved_result import ImageChunkRetrievedResultRepository
-from autorag_research.orm.schema import ImageChunkRetrievedResult
+from autorag_research.orm.schema import ImageChunk, ImageChunkRetrievedResult, Pipeline, Query
 
 
 @pytest.fixture
@@ -59,10 +59,34 @@ def test_bulk_insert(
     image_chunk_retrieved_result_repository: ImageChunkRetrievedResultRepository,
     db_session: Session,
 ):
-    inserted_keys = {(2, 2, 2), (4, 2, 1)}
+    test_queries = [Query(contents="bulk insert query 1"), Query(contents="bulk insert query 2")]
+    test_pipeline = Pipeline(name="bulk_insert_pipeline", config={"type": "test"})
+    test_image_chunks = [
+        ImageChunk(contents=b"chunk-1", mimetype="image/png"),
+        ImageChunk(contents=b"chunk-2", mimetype="image/png"),
+    ]
+    db_session.add_all([*test_queries, test_pipeline, *test_image_chunks])
+    db_session.flush()
+
+    query_ids = [query.id for query in test_queries]
+    pipeline_id = test_pipeline.id
+    inserted_keys = {
+        (query_ids[0], pipeline_id, test_image_chunks[0].id),
+        (query_ids[1], pipeline_id, test_image_chunks[1].id),
+    }
     results_to_insert = [
-        {"query_id": 4, "pipeline_id": 2, "image_chunk_id": 1, "rel_score": 0.75},
-        {"query_id": 2, "pipeline_id": 2, "image_chunk_id": 2, "rel_score": 0.65},
+        {
+            "query_id": query_ids[0],
+            "pipeline_id": pipeline_id,
+            "image_chunk_id": test_image_chunks[0].id,
+            "rel_score": 0.75,
+        },
+        {
+            "query_id": query_ids[1],
+            "pipeline_id": pipeline_id,
+            "image_chunk_id": test_image_chunks[1].id,
+            "rel_score": 0.65,
+        },
     ]
 
     inserted_count = image_chunk_retrieved_result_repository.bulk_insert(results_to_insert)
@@ -70,14 +94,19 @@ def test_bulk_insert(
 
     assert inserted_count == 2
 
-    results = image_chunk_retrieved_result_repository.get_by_query_and_pipeline([2, 4], 2)
-    assert len(results) == 3
-    assert {(result.query_id, result.pipeline_id, result.image_chunk_id) for result in results} == inserted_keys | {
-        (4, 2, 2)
-    }
+    results = image_chunk_retrieved_result_repository.get_by_query_and_pipeline(query_ids, pipeline_id)
+    result_keys = {(result.query_id, result.pipeline_id, result.image_chunk_id) for result in results}
+
+    assert len(results) == len(inserted_keys)
+    assert result_keys == inserted_keys
 
     # Cleanup
     for result in results:
         if (result.query_id, result.pipeline_id, result.image_chunk_id) in inserted_keys:
             db_session.delete(result)
+    for image_chunk in test_image_chunks:
+        db_session.delete(image_chunk)
+    db_session.delete(test_pipeline)
+    for query in test_queries:
+        db_session.delete(query)
     db_session.commit()

--- a/tests/autorag_research/pipelines/generation/test_self_rag_pipeline.py
+++ b/tests/autorag_research/pipelines/generation/test_self_rag_pipeline.py
@@ -87,6 +87,23 @@ class TestSelfRAGPipelineBehavior:
             "critique": "Need revision",
         }
 
+    def test_parse_reflection_accepts_key_value_aliases(self):
+        """Self-RAG key-value fallback parsing should honor should_retrieve/is_supported aliases."""
+        from autorag_research.pipelines.generation.self_rag import SelfRAGPipeline
+
+        pipeline = SelfRAGPipeline.__new__(SelfRAGPipeline)
+
+        reflection = pipeline._parse_reflection(
+            "should_retrieve: true\nis_supported: false\nfollow_up_query: largest planet\ncritique: need evidence"
+        )
+
+        assert reflection == {
+            "action": "RETRIEVE",
+            "supported": False,
+            "search_query": "largest planet",
+            "critique": "need evidence",
+        }
+
     @pytest.mark.asyncio
     async def test_generate_skips_retrieval_when_initial_answer_is_supported(
         self, session_factory, cleanup, mock_retrieval_pipeline

--- a/tests/autorag_research/pipelines/generation/test_self_rag_pipeline.py
+++ b/tests/autorag_research/pipelines/generation/test_self_rag_pipeline.py
@@ -70,6 +70,23 @@ class TestSelfRAGPipelineConfig:
 
 
 class TestSelfRAGPipelineBehavior:
+    def test_parse_reflection_treats_string_booleans_as_booleans(self):
+        """Self-RAG JSON reflection parsing should honor quoted true/false values."""
+        from autorag_research.pipelines.generation.self_rag import SelfRAGPipeline
+
+        pipeline = SelfRAGPipeline.__new__(SelfRAGPipeline)
+
+        reflection = pipeline._parse_reflection(
+            '{"should_retrieve": "false", "is_supported": "false", "critique": "Need revision"}'
+        )
+
+        assert reflection == {
+            "action": "REVISE",
+            "supported": False,
+            "search_query": "",
+            "critique": "Need revision",
+        }
+
     @pytest.mark.asyncio
     async def test_generate_skips_retrieval_when_initial_answer_is_supported(
         self, session_factory, cleanup, mock_retrieval_pipeline

